File: OCP/C#/Products/WithViolation/ProductFilter.cs
namespace WithViolation
{
    public class ProductFilter
    {
        public IEnumerable<Product> FilterBySize(IEnumerable<Product> products, Size size)
        {
            foreach (var product in products)
                if (product.Size == size)
                    yield return product;
        }

        public IEnumerable<Product> FilterByColor(IEnumerable<Product> products, Color color)
        {
            foreach (var product in products)
                if (product.Color == color)
                    yield return product;
        }

        public IEnumerable<Product> FilterBySizeAndColor(IEnumerable<Product> products, Size size, Color color)
        {
            foreach (var product in products)
                if (product.Size == size && product.Color == color)
                    yield return product;
        }
    }
}

File: OCP/Python/Products/WithViolation/product_filter.py
from color import Color
from product import Product
from size import Size


class ProductFilter:
    def filter_by_color(self, products: list[Product], color: Color) -> Product:
        for product in products:
            if product.color == color:
                yield product

    def filter_by_size(self, products: list[Product], size: Size) -> Product:
        for product in products:
            if product.size == size:
                yield product

    def filter_by_size_and_color(self, products: list[Product], size: Size, color: Color) -> Product:
        for product in products:
            if product.size == size and product.color == color:
                yield product